File: 卡码/代码随想录/C++/动态规划/dp/03-打家劫舍/LeetCode_0337.cpp
//337. 打家劫舍 III
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
	vector<int> traversal(TreeNode* root){
		if(!root) return vector<int>{0,0};
		vector<int> left=traversal(root->left);
		vector<int> right=traversal(root->right);
		//不偷
		int v0=max(left[0],left[1])+max(right[0],right[1]);
		//偷
		int v1=root->val+left[0]+right[0];
		return {v0,v1};
	}
	int rob(TreeNode* root) {
		vector<int> res=traversal(root);
		return max(res[0],res[1]);
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/动态规划/_03_打家劫舍/LeetCode_0337.java
package com.kurumi.codeThoughts.动态规划._03_打家劫舍;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 337. 打家劫舍 III
 */
public class LeetCode_0337 {
    static class Solution {
        public int[] traversal(TreeNode root) {
            if (root == null)
                return new int[] { 0, 0 };
            int[] left = traversal(root.left);
            int[] right = traversal(root.right);
            // 不偷
            int v0 = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
            // 偷
            int v1 = left[0] + right[0] + root.val;
            return new int[] { v0, v1 };
        }

        public int rob(TreeNode root) {
            int[] res = traversal(root);
            return Math.max(res[0], res[1]);
        }
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
